cloud-controller-manager/do/loadbalancers.go

```diff
@@ -928,6 +939,14 @@ func buildTLSForwardingRule(forwardingRule *godo.ForwardingRule, service *v1.Ser
 		return errors.New("either certificate id should be set or tls pass through enabled, not both")
 	}
 
+	var targetProtocol string
+	if _, err := getProtocol(service, annDOTargetProtocol, func(protocol string) { targetProtocol = protocol }); err != nil {
```

Comment:
Passing in an anonymous function seems a bit too complicated to me; and I'm still thinking about whether the new logic would do the right thing under all circumstances.

How about we design it like this (untested):

```golang
function getEntryProtocol(service *v1.Service) (string, error) {
  protocol, err := getProtocol(service, annDOProtocol, supportedEntryProtocols)
  if err != nil {
    return "", err
  }

  if protocol == "" {
    protocol = protoTCP
  }
  
  return protocol, nil
}

function getTargetProtocol(service *v1.Service) (string, error) {
  protocol, err := getProtocol(service, annDOTargetProtocol, supportedTargetProtocols)
  if err != nil {
    return "", err
  }

  if protocol == "" {
    protocol = protoHTTP
  }

  return protocol, nil
}

func getProtocol(service *v1.Service, annotation string, supportedProtocols []string) (string, error) {
    protocol, ok := service.Annotations[annotation]
    if !ok {
      return "", nil
    }

    if !slices.Contains(supportedProtocols, protocol) {
      return "", fmt.Errorf("invalid protocol %q specified in annotation %q", protocol, annotation)
    }

    return protocol, nil
}
```

WDYT?

Reply:
Addressed in 79364406117831963aa9814fd3fcc47b212f5476.